### PR TITLE
Distinguish between objects and directories when synchronizing filesystems

### PR DESCRIPTION
**Pull request by forgems**
#### This change allows to synchronize between two bucket based cloud storages that use key/value storage

This change adds sorting by object type to DirEntries sorting. Additionally it distinguishes between objects and directories in matchListings function. By adding this directories with the same name as object name are  distinguished from objects, and can be synchronized. This allows to synchronize files between cloud storages that are key/value stores. For example when the source has the following structure (taken with rclone ls -R --files-only)
```
index
index/test
index/test2
```
previously it couldn't synchronize the `index` file because it had the same name as directory `index/`. After this change the `index` file and `index` directory are distinguished and are not reported as duplicates. This could be a problem when you are synchronizing the files from Key/Value storage to local filesystem, because it doesn't support duplicate directory entries, and right now it reports the error about it. 

#### Was the change discussed in an issue or in the forum before?
The relevant discussion is here
https://forum.rclone.org/t/problem-when-cloning-one-cloud-storage-to-another-cloud-storage/10146/5

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: fs/direntries_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	"io"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/ncw/rclone/fs/hash"
+	"github.com/stretchr/testify/assert"
+)
+
+type FakeObject string
```
Comment (ncw) [inline fs/direntries_test.go]: You should be able to use `fstest/mockobject` here I think rather than build your own FakeObject

Comment (ncw) [reply fs/direntries_test.go]: If it is giving you a circular import then change the package to `fs_test`

**Comment by ncw** [comment]
Can you have a go at fixing up the travis output?  This seems to have broken the `TestSyncUTFNorm` test.

File: fs/direntries_test.go (hunk 1 above)
Comment (forgems) [reply fs/direntries_test.go]: Yes. I've got circular imports when I tried to use mockobject and mockdir. 
Thanks I'll change it.

Comment (forgems) [reply fs/direntries_test.go]: Fixed

**Comment by forgems** [comment]
Hi @ncw.  I've fixed the issue with `TestSyncUTFNorm`. Unfortunately, it resulted in some code duplication in `matchListings` (I couldn't reuse `CompareDirEntries` function) because it compares transformed names from `matchEntry` instead of remote names.

**Comment by ncw** [comment]
I think this looks great now - I'll merge now.

Apologies for the delay!

**Pull request merged**